### DISCUSSION TITLE
i have proplem with run more than 1 middleware

### DISCUSSION THREAD

**Discussion by Eng1Mahmoud** [vercel/next.js#71898] [Help] [answered]
### Summary

only localization Middleware runs, how can I fix this . i want to run all Middleware , first for internationalization and then auth and then etc ,

my proplem only first middleware run and other not run because first middleware return a response 

### Additional information

```js
you can see my problem from here https://github.com/amannn/next-intl/discussions/1455
```

**Comment by icyJoseph** [comment]
I think, necessarily, you need to enhance your middleware functions, to take an optional response object. 

In the next-intl docs, they have this section, where they interface with supabase:

- https://next-intl-docs.vercel.app/docs/routing/middleware#example-integrating-with-supabase-authentication

So here, https://github.com/Eng1Mahmoud/watch-store/blob/localization/middleware/authMiddleware.ts, you need to extend your "middleware" function, to take a second argument, which is the potential response given by a possible previous step.

If there is such a response, return that when it is not restricted access, otherwise, `.next()` (although, returning undefined is ok too)

There are other approaches, but I think, the least cost, is for you to do this small tweak.

**Reply by icyJoseph** [reply, marked as answer]
@Eng1Mahmoud 

I think, you can do this (I am free-styling this):

```ts
async function runMiddlewares(request: NextRequest, middlewares: Function[]) {
  let response: Response;

  for (const middleware of middlewares) {
    response = await middleware(request, response);

    if (response && !response.ok) {
      // the middleware gave a response, and it was not ok (not in 200-299 range), most likely a redirect
      return response;
    }
  }

  // return the last response
  return response;
}
```

And then modify your auth middleware to:

```ts
import { NextResponse } from "next/server";
import type { NextRequest } from "next/server";
import { decodeToken } from "@/utils/decodeToken";
import { routes } from "./helper/RolesAndRoutes";
import type { Role } from "./helper/RolesAndRoutes";
export function authMiddleware(request: NextRequest, response: NextResponse) {
  const { pathname } = request.nextUrl;
  console.log("pathname.......................", pathname);
  
  // other stuff that does redirect

  if (!route) {
    return response || NextResponse.next(); // Allow access if route is not restricted
  }
  
  // some more stuff that does redirect

  return response || NextResponse.next(); // Allow access if role is authorized
}
```

**Reply by icyJoseph** [reply]
This one would override a potential rewrite though, or data responses 🤔, maybe you'd need to also add checks for:

- response.body in the response, 
- and for the presence of the x-middleware-rewrite header ~

**Reply by Eng1Mahmoud** [reply]
It seems like it's complicated from next-intl middleware side,  i check it before  i post this issue here,  i need return  response  to make every thing work fine in next-intl @icyJoseph

**Reply by icyJoseph** [reply]
Yeah, but with those changes you'd do that, at least as you have it (with only two Middleware) because it passes along the response. The thing is that the Middleware you write, has to always pass along, or generate a response.

**Reply by Eng1Mahmoud** [reply]
Thanks Dear @icyJoseph  I tryed this solution and it work fine now I hope continue work not other things occur 😅

**Reply by icyJoseph** [reply]
No worries, always here to help and support! Don't forget to mark as answered. And by all means, come back if things get more complicated, or complex, we'll take a look at it.

**Reply by Eng1Mahmoud** [reply]
Thanks @icyJoseph

**Comment by aniruddhaadak80** [comment]
Hi @Eng1Mahmoud ,

To ensure multiple middlewares run in your Next.js application, you can follow these steps:

1. **Modify Middleware Logic**: Ensure that your first middleware does not return a response immediately. Instead, it should call `next()` to pass control to the next middleware. For example:

   ```javascript
   export function middleware(req, res, next) {
     // Your localization logic here
     
     // Call next middleware
     next();
   }
   ```

2. **Use `next()` Function**: Each middleware should call the `next()` function to proceed to the subsequent middleware. If a middleware does return a response, subsequent middlewares will not execute.

3. **Order of Middleware**: Make sure to declare your middlewares in the correct order in your application, especially if they depend on each other (e.g., localization before authentication).

4. **Check Middleware Configuration**: Verify that your middleware is correctly set up in the `middleware.ts` or `middleware.js` file and that they are exported properly.

By following these guidelines, you should be able to run all of your middlewares as intended. Thank you!

**Comment by Eng1Mahmoud** [comment]
How do I do that? I want help writing the correct code to implement all
middelware correctly @icyJoseph

**Reply by aniruddhaadak80** [reply]
Hi @Eng1Mahmoud ,

To ensure that all your middlewares run, you need to make sure that each middleware calls the `next()` function unless it is sending a response. If your first middleware is returning a response, the execution will stop, and subsequent middlewares won't run.

Here's a general approach to modify your middleware:

```javascript
const localizationMiddleware = (req, res, next) => {
    // Logic for localization
    if (someCondition) {
        return res.status(400).send('Localization error'); // Change this to `next()` if you want to continue
    }
    next(); // Call next() to proceed to the next middleware
};

const authMiddleware = (req, res, next) => {
    // Logic for authentication
    if (!req.user) {
        return res.status(401).send('Unauthorized'); // Send response if not authenticated
    }
    next(); // Proceed to the next middleware
};

// Add your other middlewares similarly...

app.use(localizationMiddleware);
app.use(authMiddleware);
// Add more middlewares as needed...

app.get('/', (req, res) => {
    res.send('Hello World!');
});
```

In this structure, each middleware must call `next()` unless it's sending a response. This allows your application to proceed through all the defined middlewares.

Feel free to adjust the logic in your middlewares to suit your needs!

Thanks!

**Reply by icyJoseph** [reply]
These look more like LLM, ChatGpt and what not answers. It is confused with express like - Middleware and Next.js Middleware. 

The right approach is as I linked in my comment above.

**Reply by Eng1Mahmoud** [reply]
@aniruddhaadak80 what you referr it it work with nodejs and express not with nextjs middleware

**Reply by aniruddhaadak80** [reply]
Hey @Eng1Mahmoud , 

Example of Multiple Midleware in App Router
Here’s an example above  :
```
// middleware.js

import { NextResponse } from 'next/server';

// Localization Middleware
export function middleware(req) {
    const localizationResult = localizationMiddleware(req);
    
    if (localizationResult) {
        return NextResponse.json(localizationResult); // Only respond if necessary
    }

    // Continue to the next middleware
    return NextResponse.next();
}

// Sample localization middleware function
function localizationMiddleware(req) {
    // Your logic for localization
    // Return localization data if applicable, otherwise return null/undefined
    const lang = req.headers.get('accept-language') || 'en';
    // Example logic: if a specific header is missing, return a response
    if (!lang) {
        return { error: "Language not specified" };
    }
    return null; // No issues, continue
}

// Authentication Middleware (could be another middleware file or inline)
export function authMiddleware(req) {
    const token = req.cookies.get('token');
    
    if (!token) {
        return NextResponse.redirect('/login'); // Redirect if not authenticated
    }
    return NextResponse.next(); // Proceed if authenticated
}

// Combining Middleware
export function combinedMiddleware(req) {
    const localizationResult = localizationMiddleware(req);
    
    if (localizationResult) {
        return NextResponse.json(localizationResult);
    }

    const authResult = authMiddleware(req);
    if (authResult) {
        return authResult; // Return redirect response if authentication fails
    }

    return NextResponse.next(); // If everything is fine, proceed
}
```

Try this out . It might help .

Thanks .
